Honor start/end in the remote read request hints

pracucci opened the pull request:
#### What this PR does

Currently, Mimir ignored remote read requests hints. When Prometheus issue a remote read request, the time range in hints may be different than the start/end time range. For example, if you run `count_over_time(up[1h]) + count_over_time(up[2h])` (example query) in Prometheus you get 2 remote read requests with the following time ranges:

```
start: 1718859107525 (04:51:47) end: 1718869907525 (07:51:47) hints start: 1718862707525 (05:51:47) hints end: 1718869907525 (07:51:47)
start: 1718859107525 (04:51:47) end: 1718869907525 (07:51:47) hints start: 1718859107525 (04:51:47) hints end: 1718869907525 (07:51:47)
```

_Apparently, the start/end is always the max one, and hints may specify a smaller one (apparently, it looks the opposite than a normal query execution where start/end should be the queried time range and hints the actual time range for which you want the raw data to process in PromQL)._

Anyway, read hints are expected to be honored. Prometheus, which implements remote read endpoints too, honor the hints too.

In this PR I'm proposing to honoring read hints, but adding an extra protection for zero-valued hint values. The reason is that we don't know if in the while there are clients passing zero-valued hints (by mistake) and I would like to protect from it.

#### Which issue(s) this PR fixes or relates to

N/A

#### Checklist

- [ ] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

krajorama commented on CHANGELOG.md:
```diff
@@ -70,6 +70,7 @@
 * [ENHANCEMENT] Query-frontend: added `remote_read` to `op` supported label values for the `cortex_query_frontend_queries_total` metric. #8412
 * [ENHANCEMENT] Query-frontend: log the overall length and start, end time offset from current time for remote read requests. The start and end times are calculated as the miminum and maximum times of the individual queries in the remote read request. #8404
 * [ENHANCEMENT] Storage Provider: Added option `-<prefix>.s3.dualstack-enabled` that allows disabling S3 client from resolving AWS S3 endpoint into dual-stack IPv4/IPv6 endpoint. Defaults to true. #8405
+* [ENHANCEMENT] Querier: honor the start/end time range specified in the read hints when executing a remote read request. #8431
```
Have you considered it `[CHANGE]` ? Technically the result might be different in some cases.  And I think we should call this out more in the release notes.

krajorama reviewed (commented):
I think the only thing missing to make the story whole is to have e2e tests try it. 
So enhance the interfaces [RemoteRead](https://github.com/grafana/mimir/blob/cf32b36884feced481c3e9127783fd243337a04e/integration/e2emimir/client.go#L259) and [RemoteReadChunks](https://github.com/grafana/mimir/blob/cf32b36884feced481c3e9127783fd243337a04e/integration/e2emimir/client.go#L303C18-L303C34) be able to take an optional `*prompb.ReadHints`.

pracucci commented:
> I think the only thing missing to make the story whole is to have e2e tests try it. So enhance the interfaces [RemoteRead](https://github.com/grafana/mimir/blob/cf32b36884feced481c3e9127783fd243337a04e/integration/e2emimir/client.go#L259) and [RemoteReadChunks](https://github.com/grafana/mimir/blob/cf32b36884feced481c3e9127783fd243337a04e/integration/e2emimir/client.go#L303C18-L303C34) be able to take an optional `*prompb.ReadHints`.

Integration tests were a bit tricky. The most complicated thing is that when you query chunks, the returned chunks are not cut to the queried time range but we return full chunks with samples within the queried time range. I've refactored `TestQuerierStreamingRemoteRead` and added a test case to query a subrange via hints.

pracucci commented on integration/querier_remote_read_test.go:
```diff
@@ -59,164 +63,258 @@ func TestQuerierRemoteRead(t *testing.T) {
+					EndMs:   now.Add(10 * time.Second).UnixMilli(),
+				},
 			},
+			// Mimir doesn't cut returned chunks to the requested time range for performance reasons. This means
+			// that we get the entire chunks in output. TSDB targets to cut chunks once every 120 samples, but
+			// it's based on estimation and math is not super accurate. That's why we get these not-perfectly-aligned
+			// chunk ranges. However, they're consistent, so tests are stable.
+			expectedStartMs: now.Add(-1500 * time.Millisecond).UnixMilli(),
+			expectedEndMs:   now.Add(10100 * time.Millisecond).UnixMilli(),
```
Note to reviewers: this is due to how TSDB `computeChunkEndTime()` works. I've confirmed it doing the math manually.

pracucci commented on integration/querier_remote_read_test.go:
```diff
@@ -33,6 +33,10 @@ func TestQuerierRemoteRead(t *testing.T) {
 		BlocksStorageS3Flags(),
+		map[string]string{
+			// This test writes samples sparse in time. We don't want compaction to trigger while testing.
+			"-blocks-storage.tsdb.block-ranges-period": "2h",
```
Note to reviewers: the default in integration tests is 1m, but I don't want such a small block period range in these tests. Reason is that chunks are also cut to period ranges, making the assertion even more complicated.

pracucci commented on integration/querier_remote_read_test.go:
```diff
@@ -59,164 +63,258 @@ func TestQuerierRemoteRead(t *testing.T) {
-			// Start dependencies.
-			minio := e2edb.NewMinio(9000, blocksBucketName)
+		// This test writes samples sparse in time. We don't want compaction to trigger while testing.
+		"-blocks-storage.tsdb.block-ranges-period": "2h",
```
Note to reviewers: the default in integration tests is 1m, but I don't want such a small block period range in these tests. Reason is that chunks are also cut to period ranges, making the assertion even more complicated.

krajorama reviewed (approved):
LGTM

pracucci commented:
>> I think the only thing missing to make the story whole is to have e2e tests try it. So enhance the interfaces [RemoteRead](https://github.com/grafana/mimir/blob/cf32b36884feced481c3e9127783fd243337a04e/integration/e2emimir/client.go#L259) and [RemoteReadChunks](https://github.com/grafana/mimir/blob/cf32b36884feced481c3e9127783fd243337a04e/integration/e2emimir/client.go#L303C18-L303C34) be able to take an optional *prompb.ReadHints.
> 
> Integration tests were a bit tricky. The most complicated thing is that when you query chunks, the returned chunks are not cut to the queried time range but we return full chunks with samples within the queried time range. I've refactored TestQuerierStreamingRemoteRead and added a test case to query a subrange via hints.

I reverted back `TestQuerierRemoteRead` to the `main` branch version one in https://github.com/grafana/mimir/pull/8431/commits/a10436dae79c1e2892ed6eda8f1d6dc664d7be0d. Unfortunately the one in main branch is broken (assertions are skipped!) but I need more time to fix it, so I will do in a separate PR. In the meanwhile I merge this PR to unlock this change.
